File: 139-word-break/word-break.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        if s in wordDict:
            return True
        dp=[True]+[False]*len(s)
        sub_str=s
        for i in range(1,len(s)+1):
            for word in wordDict:
                start=i-len(word)
                if start>=0 and dp[start] and s[start:i]==word:
                    dp[i]=True
                    break

        return dp[-1]
        

File: 139-word-break/word-break.cpp
class Solution {
public:
    bool solve(string& s,unordered_set<string>& st,int start,vector<int>& dp){
        if(start==(int)s.size()) return true;
        if(dp[start]!=-1) return dp[start];
        string match="";
        for(int i=start;i<s.size();i++){
            match.push_back(s[i]);
            if(st.count(match)){
                if(solve(s,st,i+1,dp)) return dp[start]=true;
            }
        }
        return dp[start]=false;
    }
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string>st(wordDict.begin(),wordDict.end());
        int n=s.size();
        vector<int> dp(n+1,-1);
        // dp[0]=true;
        return solve(s,st,0,dp);
    }
};